src/Sentry/Internal/DefaultSentryStructuredLogger.cs

```diff
@@ -0,0 +1,131 @@
+
+namespace Sentry.Internal;
+
+internal sealed class DefaultSentryStructuredLogger : SentryStructuredLogger
+{
+    private readonly IHub _hub;
+    private readonly IInternalScopeManager _scopeManager;
+    private readonly SentryOptions _options;
+    private readonly ISystemClock _clock;
+
+    private readonly bool _isEnabled;
+
+    internal DefaultSentryStructuredLogger(IHub hub, IInternalScopeManager scopeManager, SentryOptions options, ISystemClock clock)
+    {
+        _hub = hub;
+        _scopeManager = scopeManager;
+        _options = options;
+        _clock = clock;
+
+        _isEnabled = options is { Experimental.EnableLogs: true };
+    }
+
+    private protected override void CaptureLog(SentryLogLevel level, string template, object[]? parameters, Action<SentryLog>? configureLog)
+    {
+        if (!_isEnabled)
+        {
+            return;
+        }
+
+        var timestamp = _clock.GetUtcNow();
+
+        if (!TryGetTraceId(_hub, _scopeManager, out var traceId))
+        {
+            _options.DiagnosticLogger?.LogWarning("TraceId not found");
+        }
+
+        _ = TryGetParentSpanId(_hub, _scopeManager, out var parentSpanId);
+
+        string message;
+        try
+        {
+            message = string.Format(CultureInfo.InvariantCulture, template, parameters ?? []);
+        }
+        catch (FormatException e)
+        {
+            _options.DiagnosticLogger?.LogError(e, "Template string does not match the provided argument. The Log will be dropped.");
+            return;
+        }
+
+        SentryLog log = new(timestamp, traceId, level, message)
+        {
+            Template = template,
+            Parameters = ImmutableArray.Create(parameters),
+            ParentSpanId = parentSpanId,
+        };
+
+        try
+        {
+            configureLog?.Invoke(log);
+        }
+        catch (Exception e)
+        {
+            _options.DiagnosticLogger?.LogError(e, "The configureLog callback threw an exception. The Log will be dropped.");
+            return;
+        }
+
+        log.SetAttributes(_options);
+
+        var configuredLog = log;
+        if (_options.Experimental.BeforeSendLogInternal is { } beforeSendLog)
+        {
+            try
+            {
+                configuredLog = beforeSendLog.Invoke(log);
+            }
+            catch (Exception e)
+            {
+                _options.DiagnosticLogger?.LogError(e, "The BeforeSendLog callback threw an exception. The Log will be dropped.");
+                return;
+            }
+        }
+
+        if (configuredLog is not null)
+        {
+            //TODO: enqueue in Batch-Processor / Background-Worker
+            // see https://github.com/getsentry/sentry-dotnet/issues/4132
+            _ = _hub.CaptureEnvelope(Envelope.FromLog(configuredLog));
+        }
+    }
+
+    private static bool TryGetTraceId(IHub hub, IInternalScopeManager? scopeManager, out SentryId traceId)
+    {
+        if (hub.GetSpan() is { } span)
```

Comment:
We also have `hub.Span` which takes precedence I believe?
The logic to extract the trace-id needs to stay in a single place somehow so we avoid different behavior throughout the SDK

Reply:
There's no `hub.Span`... there's a `Scope.Span` though (which is what this current code is implicitly reading):
https://github.com/getsentry/sentry-dotnet/blob/6e50eaa87b3aa24a921824c4494566b01f0640f0/src/Sentry/Internal/Hub.cs#L244

And [that prioritises the internal _span property](https://github.com/getsentry/sentry-dotnet/blob/34d849038e3e23868a12d56ea256c36b0a7f3a61/src/Sentry/Scope.cs#L617-L629) that you're talking about @bruno-garcia.

Reply:
thanks for the hint
replaced with `Hub.GetTraceHeader()`